runtime/vm/JVMImage.cpp

```diff
@@ -147,58 +151,70 @@ JVMImage::readImageFromFile()
 {
 	OMRPORT_ACCESS_FROM_OMRPORT(getPortLibrary());
 
-	// TODO: The size will need to be very large or dynamically based on whatever is written to the dump
-	// TODO: Fetch the initial size of the jvm image
-    char imageBuffer[JVMImage::INITIAL_IMAGE_SIZE];
-    memset(imageBuffer, 0, sizeof(imageBuffer));
+	omrthread_monitor_enter(_jvmImageMonitor);
 
 	intptr_t fileDescriptor = omrfile_open(_dumpFileName, EsOpenRead, 0444);
 	if (fileDescriptor == -1) {
 		return false;
 	}
 
-	intptr_t bytesRead = omrfile_read(fileDescriptor, imageBuffer, sizeof(imageBuffer));
-	if (bytesRead == -1) {
+	// Read image header then mmap the rest of the image (heap) into memory
+	JVMImageHeader imageHeader;
+	omrfile_read(fileDescriptor, &imageHeader, sizeof(JVMImageHeader));
+	uint64_t fileSize = omrfile_flength(fileDescriptor);
+	if (fileSize != sizeof(JVMImageHeader) + imageHeader.imageSize) {
 		return false;
 	}
 
-	if (omrfile_close(fileDescriptor) != 0) {
-		return false;
-	}
+	// Mmap our file into virtual memory
+	void *block = mmap((void*)imageHeader.heapAddress, sizeof(JVMImageHeader) + imageHeader.imageSize, PROT_READ, MAP_PRIVATE, fileDescriptor, 0);
+	_heap = (J9Heap*)((char*)block + sizeof(JVMImageHeader));
+
+	omrfile_close(fileDescriptor);
 
-	// TODO: Finally load the JVMImage using the data we read from the image
+	omrthread_monitor_exit(_jvmImageMonitor);
 
 	return true;
 }
 
-void
+bool
 JVMImage::storeImageInFile()
 {
 	OMRPORT_ACCESS_FROM_OMRPORT(getPortLibrary());
 
+	omrthread_monitor_enter(_jvmImageMonitor);
```

Comment:
We might need this but I am not sure if any threads would be changing heap when we call storeImageInFile.

Reply:
Because we would only be calling storeImageInFile at shutdown (after all allocation is done)